### PR TITLE
platformio: 3.4.1 -> 3.5.0

### PR DESCRIPTION
**Pull request by mogorman**
###### Motivation for this change
i wanted to use an esp32 so i needed to update platformio. it needs a specific version of lru_cache because arrow 0.12.0 needs a specific version it seems. also the wrapper function had to be changed as it was not linking to the lru package correctly.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/arrow/default.nix [hunk 1]
```diff
@@ -16,7 +16,7 @@ buildPythonPackage rec {
   '';
 
   checkInputs = [ nose chai simplejson ];
-  propagatedBuildInputs = [ dateutil backports_functools_lru_cache ];
+  propagatedBuildInputs = [ dateutil backports_functools_lru_cache_121 ];
```
Comment (adisbladis) [inline pkgs/development/python-modules/arrow/default.nix]: I would rather patch the line `install_requires.append('backports.functools_lru_cache==1.2.1')` from `setup.py` to read `install_requires.append('backports.functools_lru_cache>=1.2.1')`.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -1007,6 +1007,8 @@ in {
 
   backports_functools_lru_cache = callPackage ../development/python-modules/backports_functools_lru_cache { };
 
+  backports_functools_lru_cache_121 = callPackage ../development/python-modules/backports_functools_lru_cache_121 { };
```
Comment (adisbladis) [inline pkgs/top-level/python-packages.nix]: Even if `arrow` really does require a very specific version of `lru_cache` this version should not be exposed in `python-packages.nix`.

Ideally `arrow` should be patched to not require a very explicit version but if it really does this override should be moved to the `arrow` directory.

**Review by adisbladis — CHANGES_REQUESTED** [review]

File: pkgs/top-level/python-packages.nix (hunk 2 above)
Comment (FRidh) [reply pkgs/top-level/python-packages.nix]: Additional versions of packages are not permitted in `python-packages.nix` *at all* as they cause collisions, so also not inside expressions. Applications that are defined *outside* `python-packages.nix` can override the package set to insert whatever version they need.

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/arrow/default.nix (hunk 1 above)
Comment (FRidh) [reply pkgs/development/python-modules/arrow/default.nix]: I've patched the `setup.py` in 340cadd97d9.

**Comment by mogorman** [comment]
mad changes requested by @FRidh and @adisbladis and verified it still worked on my box

File: pkgs/development/arduino/platformio/chrootenv.nix [hunk 3]
```diff
@@ -30,5 +32,9 @@ buildFHSUserEnv {
     platforms = with platforms; linux;
   };
 
+  extraInstallCommands = ''
+    ln -s $out/bin/platformio $out/bin/pio
+  ''a
```
Comment (FRidh) [inline pkgs/development/arduino/platformio/chrootenv.nix]: typo

File: pkgs/development/python-modules/platformio/default.nix [hunk 4]
```diff
@@ -23,6 +23,13 @@ buildPythonPackage rec {
 
   patches = [ ./fix-searchpath.patch ];
 
+  postInstall = ''
```
Comment (FRidh) [inline pkgs/development/python-modules/platformio/default.nix]: Why is this needed? `wrapPythonPrograms` already sets `PATH` and injects `PYTHONPATH`.
Note also that the `postFixup` (or `fixupPhase` in general) is preferred for wrappers.

Comment (mogorman) [reply pkgs/development/python-modules/platformio/default.nix]: you are right its no longer needed with the other fixes in place. earlier i was not seeing the lru getting linked in properly.

File: pkgs/development/arduino/platformio/chrootenv.nix (hunk 3 above)
Comment (mogorman) [reply pkgs/development/arduino/platformio/chrootenv.nix]: fixed

**Comment by FRidh** [comment]
What are the packages in the chrootenv for? They all seem to be dependencies of platformio already. Did you get an import error if you did not add say the backports package?

**Comment by mogorman** [comment]
Earlier i  thought I did. I just tested it with them removed and it runs fine as well. I pushed updated code just now @FRidh

**Review by adisbladis — APPROVED** [review]

**Comment by FRidh** [comment]
I think you may as well write `python.withPackages(ps: with ps; [ platformio ])` instead of the current dependencies.

**Comment by mogorman** [comment]
@FRidh i am happy to make that change but im not sure how. i looked at some of the packages that make use of it but not sure i completely grok whats going on.

**Comment by FRidh** [comment]
@mogorman it creates a tree of symbolic links including the listed packages and Python dependencies.

**Comment by mogorman** [comment]
right @FRidh , im just not sure how to correctly to use it in the expression and have the resulting fhsuserenv work. i tried saying let pythonUserEnv = python.withPackages(ps: with ps; [ platformio ]) and then including that instead of the python deps and it failed to run

**Comment by FRidh** [comment]
```
i tried saying let pythonUserEnv = python.withPackages(ps: with ps; [ platformio ]) and then including that instead of the python deps and it failed to run
```
That's a pity, I though that could work. How did it fail? With what error?

**Comment by FRidh** [comment]
In any case, that's an improvement that can be made separate from this PR.
